File: Clue/GameBoard.py
#!/usr/bin/python

class GameBoard:
    _minWidth = 0
    _minHeight = 0
    _maxWidth = 23
    _maxHeight = 25

    _board = None
    _peices = None

    @property
    startSquares = {CharacterName.missScarlet : [_minWidth, 17],
                     CharacterName.profPlum : [6, _minHeight],
                     CharacterName.colMustard: [7, _maxHeight],
                     CharacterName.mrsPeacock : [18, _minHeight],
                     CharacterName.mrsWhite : [_maxWidth, 14],
                     CharacterName.mrGreen : [_maxWidth, 9]}

    def __init__(self):

        for x in range(_maxWidth):
            for y in range(_maxHeight):
                _board[x][y] = initSquare(x, y)
        

    def initBoard(self):
        for x in range(_maxWidth):
            for y in range(_maxHeight):
                _board[x][y] = initSquare(x, y)

    def initSquare(x, y):

        valid = isValidPosition(x,y):

        return GameSquare(False, valid, None):

    def isStartSquare(x, y):
        if x == _minWidth and y == 17:
            return True
        elif x == _maxWidth:
            if y == 14 or y == 9:
                return True
        elif y == _minHeight and x == 6:
            return True
        elif y == maxHeight:
            if x == 6 or x == 7:
                return True

        return False
            
    def isValidPosition(x, y):

        #start spaces are valid
        #other edges are not
        #rooms are invalid
        #all other spaces are valid
        if isStartSquare(x,y):
            return True
        #all other edge spaces are not
        elif x == _minWdith or x == _maxWidth:
            return False
        elif y == _minHeight or y == _maxHeight:
            return False
        #Study
        elif x < 4 and y < 7:
            return False
        #Library
        elif x > 5 and x < 11 and y < 6:
            return False
        #Library 2
        elif x > 6 and x < 10 and y == 6:
            return False
        #Billiard Room
        elif x > 11 and x < 17 and y < 6:
            return False
        #Conservatory 
        elif x > 18 and y < 5:
            return False
        #Conservatory 2
        elif x > 19 and y < 6:
            return False
        #Hall
        elif x > 0 and x < 7 and y > 8 and y < 15:
            return False
        #Staircase/ Clue center
        elif x > 7 and x < 15 and y > 8 and y < 14:
            return False
        #Ball Room
        elif x > 16 and x < 23 and y > 7 and y < 9:
            return False
        #Ball Room 2
        elif x > 16 and x < _maxWidth and y > 8 and y < 14:
            return False
        #Ball Room 3
        elif x > 16 and x < 23 and y > 13 and y < 16:
            return False
        #Lounge
        elif x > _minWidth and x < 6 and y > 16 and y < _maxHeight:
            return False
        #Dinging Room
        elif x > 8 and x < 15 and y > 17 and y < 21:
            return False
        #Dining Room 2
        elif x > 8 and x < 16 and y > 20 and y < _maxHeight:
            return False
        #Kitchen
        elif x > 17 and x < _maxWidth and y > 19 and y < _maxHeight:
            return False
        else:
            return True

    def __str__(self):

        string = ''
        for x in range(_maxWidth):
            for y in range(_maxHeight):
                square = _board[x][y]

                if square.isValid():
                    string + "[]"
                else:
                    string + "{}"
            string = "\n"
        return string
    """
    def setPeices(self, peices):

        
        

    def movePiece(self, peiceName):
    """

File: Java/src/main/java/clue/logic/GameBoard.java
package clue.logic;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;
import java.lang.StringBuilder;

@Entity
public class GameBoard extends GameObject {

  public static final int minWidth = 0;
  public static final int minHeight = 0;
  public static final int maxWidth = 25;
  public static final int maxHeight = 25;

  /**
   * board is a 2d array. Here it is represented as a list
   * so that it will map better in the database. The index
   * is calculated by the following equation.
   * index = rowNumber * maxWidth + columnNumber;
   */
  @OneToMany
  List<GameSquare> board;

  @OneToMany
  List<GamePiece> pieces;

  /**
   * Constructor
   */
  public GameBoard() {
    this(new ArrayList<>());
  }

  public GameBoard(List<GamePiece> pieces) {
    label = "GameBoard";
    board = new ArrayList<>();
    for(int x = 0; x < maxWidth; ++x) {
      for(int y = 0; y < maxHeight; ++y) {
        board.add(initSquare(x, y));
      }
    }

    for (GamePiece piece: pieces) {
      getSquare(piece.getX(), piece.getY()).setPiece(piece);
    }
  }

  public GameSquare initSquare(int x, int y) {

    boolean isValid = isValidPosition(x, y);

    return new GameSquare(false, isValid, null);
  }

  public boolean isStartSquare(int x, int y) {

    if(x == minWidth && (y == 7 || y == 16)) {
      return true;
    }
    else if (x == (maxWidth - 1) && (y == 9 || y == 14)) {
      return true;
    }
    else if (y == minHeight && (x == 7 || x == 17)) {
      return true;
    }
    else if (y == (maxHeight -1) && (x == 5 || x == 18)) {
      return true;
    }
    return false;
  }

  public boolean isMiscClosedSquare(int x, int y) {

    if (x == 23 && (y == 6 || y == 17)) {
      return true;
    }
    return false;
  }

  //returns whether a given position is valid
  public boolean isValidPosition(int x, int y) {

    //start spaces are valid
    //other edges are not
    //rooms are invalid
    //all other spaces are valid
    if(isStartSquare(x, y)) {
      System.out.println("isStartSquare(" + x + ", " + y +")");
      return true;
    }
    else if (x == minWidth || x == (maxWidth - 1)) {
      return false;
    }
    else if (y == minHeight || y == (maxHeight - 1)) {
      return false;
    }
    //Lounge
    else if (x < 6 && y < 7) {
      return false;
    }
    //Dining Room 1
    else if (x > 8 && x < 15 && y < 8) {
      return false;
    }
    //Dining Room 2
    else if (x == 15 && y < 5) {
      return false;
    }
    //Kitchen
    else if (x > 17 && y < 6) {
      return false;
    }
    //Hall
    else if (x < 7 && y > 8 && y < 15) {
      return false;
    }
    //clue
    else if (x > 7 && x < 15 && y > 9 && y < 15) {
      return false;
    }
    //Ball Room 1
    else if (x > 16 && x < 23 && y > 7 && y < 16) {
      return false;
    }
    // Ball Room 2
    else if (x > 22 && y > 9 && y < 14) {
      return false;
    }
    //Study
    else if (x < 4 && y > 16) {
      return false;
    }
    //Library 1
    else if (x > 5 && x < 11 && y > 17) {
      return false;
    }
    //Library 2
    else if (x > 6 && x < 10 && y == 17) {
      return false;
    }
    //Billiard Room
    else if (x > 11 && x < 17 && y > 17) {
      return false;
    }
    //Conservatory 1
    else if (x > 18 && y > 18) {
      return false;
    }
    //Conservatory 2
    else if (x > 19 && y == 18) {
      return false;
    }
    else if (isMiscClosedSquare(x, y)) {
      return false;
    }
    else {
      return true;
    }
  }

  public List<GameSquare> getBoard() {
    return board;
  }

  public void setBoard(List<GameSquare> board) {
    this.board = board;
  }

  public List<GamePiece> getPieces() {
    return pieces;
  }

  public void setPieces(List<GamePiece> pieces) {
    this.pieces = pieces;
  }

  public GameSquare getSquare(int x, int y) {
    if((y * maxWidth + x) > board.size()) {
      System.out.println("x: " + x + " y: " + y);
      Thread.dumpStack();
    }
    return board.get(y * maxWidth + x);
  }

  public String toString() {
    GameSquare square;
    StringBuilder builder = new StringBuilder();

    for(int x = 0; x < maxWidth + 1; ++x) {
      for(int y = 0; y < maxHeight + 1; ++y) {

        if(x == maxWidth)  {
          builder.append("[" + y % 10 + "]");
          continue;
        }
        else if(y == maxHeight) {
          builder.append("[" + x % 10 + "]");
          continue;
        }

        square = getSquare(x, y);

        if(square == null) {
          System.out.println("square == null");
        }

        String squareContent = "{X}";
        if(square.isValid()) {
          if(square.isOccupied()) {
            squareContent = "[i]";
          } else {
            squareContent = "[ ]";
          }
        }
        builder.append(squareContent);
      }
      builder.append("\n");
    }

    return builder.toString();
  }
}
